chichiwang/exercise-js
# Convert server to serve universal React application

## Description
Linked to Issue: #17 
Create a universal react application such that the server serves the React application and the client script hydrates the application. Ensure all linting and tests pass. A build script should exist to build the application, and the start script should start the watcher for any changes.

## Changes
* Change the name of the `build:server` script to `build:dev`
  * The build script will build both the server and client: build scripts should be environment-based rather than package-based
* Add the following dependencies to `package.json`:
  * `react` -  Necessary for the React application
  * `react-dom` - Necessary for the server to render and serve the React application
* Add the following devDependencies to `package.json`:
  * `@babel/preset-react` - allows the transpilation of React for build and Jest
  * `@types/react` - Allows TypeScript to enforce types when interacting with React
  * `@types/react-dom` - Allows TypeScript to enforce types when interacting with React-Dom
  * `eslint-import-resolver-alias` - Allows configuration of aliases in eslint
  * `eslint-plugin-import` - Allows configuration of import rules in eslint
  * `jest-environment-jsdom` - Allows for DOM environment and corresponding globals to be avaible to Jest
  * `react-test-renderer` - Allows for snapshot testing
* Update `.estlintrc.js` configurations:
  * Allow importing `.js`, `.jsx`, `.ts`, `.tsx` files without specifying file extensions
  * Allow `console.error` and `console.warn` to exist in the client application
  * Set up aliases to match bundler
* Update `babel.config.js` configurations:
  * Allow Jest and Bundler to transpile React
* Create a React root component `client/app.tsx` to provide a Hello-World application
* Create a `client/entry.tsx` to act as an entry point to the client bundle and hydrate the server-rendered React application
* Update `jest.config.json` to map aliases to match the bundler settings
* Update `tsconfig.json` to map aliases to match the bundler settings
* Remove `tsconfig.server.json` as a server-specific configuration is not necessary at this time
* Update `webpack-config.js` to bundle both the client and server bundles and to set up aliases for the client directory for absolute import paths

## Steps to QA
* Pull down and switch to this branch
* Run `npm run build` and ensure server builds
  * Verify functionality by running `node dist/index.js` and ensure server starts. Verify in browser.
* Run `npm start` and ensure server builds and starts
  * Verify functionality in browser
  * Make a change to the app (ie: `client/App.tsx`) and ensure server rebuilds. Refresh in browser to ensure change is reflected.
